### DISCUSSION TITLE
Howto get  pck12  for wsdl

### DISCUSSION THREAD

**Discussion by jpertuz1** [Keyfactor/ejbca-ce#426] [General]
Hi. 

I have (for test) ejbca 8  ce   on Docker.  
I TEST  webservices WSDL,    The Certificate is generated without problem, but I am missing the return to be able to take the certificate. In short, the certificate is created but I can't figure out how to download it in pkcs12 (p12)

Annex part of the code.

// 
    const masterCert = await getMasterCertified();
    const agent = new https.Agent({
      rejectUnauthorized: false,
      cert: masterCert.Body,
      key: masterCert.Body,
    });
    // 
    const instance = axios.create({ httpsAgent: agent });
    const soapClient = await soap.createClientAsync(
      'https://1.1.1.1/ejbca/ejbcaws/ejbcaws?wsdl',
      {
        httpsAgent: agent,
        request: instance,
        // eslint-disable-next-line camelcase
        wsdl_options: {
          connection: 'keep-alive',
          userDataVows: 'userDataVows',
          userMatch: 'userMatch',
          extendedInformationWS: 'extendedInformationWS',
        },
      },
    );
    console.log('DOC-------', userdoc);
    // 
    const params = {
      caName: 'AC-Ap',
      endEntityProfileName: 'Entidad-SignDirect',
      certificateProfileName: 'SignerTest',
      email: user,
      extendedInformation: {
        name: 'name',
        value: 'Samuel',
      },
      username: userdoc,
      sendNotification: false,
      password: docEncrypt,
      keyRecoverable: false,
      clearPwd: false,
      tokenType: 'P12',
      subjectDN: `CN=${userdoc},O=3DLink,C=VE,EMAILADDRESS=${user}`,
      status: 10,
    };

    const rest = await soapClient.editUserAsync({ arg0: params });

    console.log('REST', rest);

    const paramsGenerate = {
      arg0: userdoc,
      arg1: userdoc,
      arg3: '2048',
      arg4: 'RSA',
    };

    const rest3 = await soapClient.pkcs12ReqAsync(paramsGenerate);

    console.log('REST 2 ', rest3);

    console.log('KEYSTORE', rest3[0].return.keystoreData);

**Comment by primetomas** [comment]
You can look at the test code for this in EJBCA. For example:
https://github.com/Keyfactor/ejbca-ce/blob/main/modules/systemtests/src-test/org/ejbca/core/protocol/ws/EjbcaWSTest.java#L838

You can also run the clientToolBox EjbcaWSRaCli and perform the commands succesfully, and if you want enable WS debugging to see the actual SOAP messages going over the wire, which might help you to understand how to get the keystore data.
https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/protocols/web-service-interface#WebServiceInterface-WebServiceMessageDebuggingWS_message_debugging

**Reply by jpertuz1** [reply]
Thanks Tomas for responding.
The detail I'm missing is being able to get the hash back to be able to grant the pkcs12 to the end user.
My entity profile is to generate a pkcs12, when I generate it by the EJBCA RA it works fine and I get it.

When I generate it through the WSDL it executes it, but what I am missing is obtaining it. At the moment my only way out is:


storage                      |       keystoreData:
'TUlBQ0FRTXdnQVlK.........XXXXX....................UE9</keystoreData></return></ns2:pkcs12ReqResponse></soap:Body></soap:Envelope>',

**Comment by primetomas** [comment]
The data you get back is the PKCS#12 data base64 encoded. You need to take that data, base64 decode it and write it to file.

**Reply by jpertuz-19** [reply]
Thanks Tomas

We solve like this:     base64 -d out-base64 | base64 -d > file.p12

Your help helped me.
